src/photon/photon_scheduler.c

```diff
@@ -303,16 +348,33 @@ void process_message(event_loop *loop,
   } break;
   case GET_TASK: {
     /* Update the task table with the completed task. */
-    if (state->db != NULL && worker->task_in_progress != NULL) {
-      task_set_state(worker->task_in_progress, TASK_STATUS_DONE);
-      task_table_update(state->db, worker->task_in_progress,
-                        (retry_info *) &photon_retry, NULL, NULL);
-      /* The call to task_table_update takes ownership of the task_in_progress,
-       * so we set the pointer to NULL so it is not used. */
-      worker->task_in_progress = NULL;
-    } else if (worker->task_in_progress != NULL) {
-      free_task(worker->task_in_progress);
-      worker->task_in_progress = NULL;
+    task *task_in_progress = worker->task_in_progress;
+    task_spec *spec = (task_in_progress != NULL)?
+        task_task_spec(task_in_progress):NULL;
+
+    /* If this worker reports a completed task: account for resources. */
+    if (task_in_progress != NULL) {
+      /* Return dynamic resources back for the task in progress. */
+      for (int i = 0 ; i < MAX_RESOURCE_INDEX; i++) {
+        state->dynamic_resources[i] += task_required_resource(spec, i);
+        /* Sanity-check resource vector boundary conditions. */
+        CHECK(state->dynamic_resources[i] <= state->static_resources[i]);
+      }
+      print_resource_info(state, spec);
+      /* If we're connected to Redis, update tables. */
+      if (state->db != NULL && task_in_progress != NULL) {
+        /* Update control state tables. */
+        task_set_state(worker->task_in_progress, TASK_STATUS_DONE);
+        task_table_update(state->db, worker->task_in_progress,
+                          (retry_info *) &photon_retry, NULL, NULL);
+        /* The call to task_table_update takes ownership of the task_in_progress,
+         * so we set the pointer to NULL so it is not used. */
+        worker->task_in_progress = NULL;
+      }
+      if (worker->task_in_progress) {
```

Comment:
the "else" here is important (we cannot free a task if ownership has been passed to task_table_update)